build-system/tasks/check-links.js

```diff
@@ -109,66 +116,75 @@ async function checkLinks() {
+ * Checks a given markdown file for dead links.
  *
- * @param {string} markdown Original markdown.
- * @return {string} Markdown after filtering out allowed links.
+ * @param {string} file
+ * @param {!Array<string>} filesWithDeadLinks
+ * @return {!Promise}
  */
-function filterAllowedLinks(markdown) {
-  let filteredMarkdown = markdown;
-
-  // localhost links optionally preceded by ( or [ (not served on Travis)
-  filteredMarkdown = filteredMarkdown.replace(
-    /(\(|\[)?http:\/\/localhost:8000/g,
-    ''
-  );
-
-  // Links in script tags (illustrative, and not always valid)
-  filteredMarkdown = filteredMarkdown.replace(/src="http.*?"/g, '');
-
-  // Links inside a <code> block (illustrative, and not always valid)
-  filteredMarkdown = filteredMarkdown.replace(/<code>([^]*?)<\/code>/g, '');
+function checkLinksInFile(file, filesWithDeadLinks) {
+  let markdown = fs.readFileSync(file).toString();
 
-  // Links inside a <pre> block (illustrative, and not always valid)
-  filteredMarkdown = filteredMarkdown.replace(/<pre>([^]*?)<\/pre>/g, '');
+  // Links inside <code> blocks are illustrative and not always valid. Must be
+  // removed because markdownLinkCheck() does not ignore them like <pre> blocks.
+  markdown = markdown.replace(/<code>([^]*?)<\/code>/g, '');
 
-  // After allow-listing is done, clean up any remaining empty blocks bounded
-  // by backticks. Otherwise, `` will be treated as the start of a code block
-  // and confuse the link extractor.
-  filteredMarkdown = filteredMarkdown.replace(/\ \`\`\ /g, '');
-
-  return filteredMarkdown;
-}
-
-/**
- * Reads the raw contents in the given markdown file, filters out localhost
- * links (because they do not resolve on Travis), and checks for dead links.
- *
- * @param {string} markdownFile Path of markdown file, relative to src root.
- * @return {Promise} Used to wait until the async link checker is done.
- */
-function runLinkChecker(markdownFile) {
-  // `.template.md` is a common suffix for files that may have interpolation
-  // tokens, possibly as part of their links. So we skip them.
-  if (path.basename(markdownFile).endsWith('.template.md')) {
-    return Promise.resolve();
-  }
-  // Skip files that were deleted by the PR.
-  if (!fs.existsSync(markdownFile)) {
-    return Promise.resolve();
-  }
-  const markdown = fs.readFileSync(markdownFile).toString();
-  const filteredMarkdown = filterAllowedLinks(markdown);
   const opts = {
-    baseUrl: 'file://' + path.dirname(path.resolve(markdownFile)),
+    // Relative links start at the markdown file's path.
+    baseUrl: 'file://' + path.dirname(path.resolve(file)),
+    ignorePatterns: [
+      // Localhost links don't work unless a `gulp` server is running.
+      {pattern: /localhost/},
+      // Templated links are merely used to generate other markdown files.
+      {pattern: /\$\{[a-z]*\}/},
+    ],
   };
-  return markdownLinkCheck(filteredMarkdown, opts);
+
+  return new Promise((resolve, reject) => {
```

Comment:
Do we need to handle promises at this level? It's possible `utils.promisify` can handle `markdownLinkCheck` just fine.

Reply:
(since we'd like to avoid `BBPromise`. using a native util for simpler code is worth it imo)

Reply:
I chose to handle promises this way in order to make the code match the API documentation at https://www.npmjs.com/package/markdown-link-check#markdownlinkcheckmarkdown-opts-callback.

Agreed though. There are several instances of `BBPromise` usage in `build-system/` that we should probably switch to `utils.promisify`. Best in a separate PR.